osmoutils/store_helper_test.go

```diff
@@ -0,0 +1,188 @@
+package osmoutils_test
+
+import (
+	"errors"
+	"fmt"
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/osmosis-labs/osmosis/v11/app/apptesting"
+	"github.com/osmosis-labs/osmosis/v11/osmoutils"
+)
+
+type TestSuite struct {
+	apptesting.KeeperTestHelper
+	store sdk.KVStore
+}
+
+const (
+	keyA               = "a"
+	keyB               = "b"
+	keyC               = "c"
+	mockStopValue      = "stop"
+	afterMockStopValue = mockStopValue + keyA
+	basePrefix         = "base"
+	prefixOne          = "one"
+	prefixTwo          = "two"
+)
+
+func TestOsmoUtilsTestSuite(t *testing.T) {
+	suite.Run(t, new(TestSuite))
+}
+
+func (s *TestSuite) SetupStoreWithBasePrefix() {
+	_, ms := s.CreateTestContextWithMultiStore()
+	prefix := sdk.NewKVStoreKey(basePrefix)
+	ms.MountStoreWithDB(prefix, sdk.StoreTypeIAVL, nil)
+	err := ms.LoadLatestVersion()
+	s.Require().NoError(err)
+	s.store = ms.GetKVStore(prefix)
+}
+
+func mockParseValue(b []byte) (string, error) {
+	return string(b), nil
+}
+
+func mockParseValueWithError(b []byte) (string, error) {
+	return "", errors.New("mock error")
+}
+
+func mockStop(b []byte) bool {
+	return string(b) == fmt.Sprintf("%s%s", prefixOne, mockStopValue)
+}
+
+func (s *TestSuite) TestGatherValuesFromIteratorWithStop() {
+
+	testcases := map[string]struct {
+		// if prefix is set, startValue and endValue are ignored.
+		// we either create an iterator prefix or a range iterator.
+		prefix     string
+		startValue string
+		endValue   string
+		preSetKeys []string
+		isReverse  bool
+
+		expectedValues []string
+		expectedErr    bool
+	}{
+		"prefix iterator, no stop": {
+			preSetKeys: []string{prefixOne + keyA, prefixOne + keyB, prefixOne + keyC},
+
+			prefix: prefixOne,
+
+			expectedValues: []string{"0", "1", "2"},
```

Comment:
Hi @p0mvn I'm trying to understand how all the iterators work, how are figure out these expected values? Would it the expected value be different if this test case was a `ReverseIterator` or just a regular `Iterator` ?

Reply:
Hi @stackman27. For every key in `preSetKeys`, we set a value corresponding to its index:
https://github.com/osmosis-labs/osmosis/blob/e5c75ce586dd7861c89c865fee58f013eab88ca6/osmoutils/store_helper_test.go#L142-L144

So, in this case `prefixOne + keyA` -> `0`, `prefixOne + keyB` -> `1`, `prefixOne + keyC` -> `2`

Yes, if it were a reverse iterator, the order would go 2, 1, 0

However, say we had the following insertion order:
- `prefixOne + keyB`
- `prefixOne + keyC`
- `prefixOne + keyA`

Then, the mapping would be
- B -> 0
- C -> 1
- A -> 2

and the expected values are:
```
[]string{"2", "0", "1"}
```

Because iterators use lexicographic order of keys where A < B < C

Let me know if that makes sense

Reply:
aah i see that makes a lot of sense! Thank you :))